### PR TITLE
[CPU] Fix std_mean/var_mean all-reduce ~4x slower than std + mean (#122191)

### LINKED ISSUE #122191: torch.std_mean slower than separate torch.mean and torch.std calls on CPU
### 🐛 Describe the bug

On cuda, `torch.std_mean` is faster than separate calls of `torch.mean` and `torch.std`. Interestingly the opposite is the case on CPU, and that is quite counterintuitive. This can lead to situations where the user needs to first check the device and then run different function calls to optimize performance across devices.

Here is a minimal script that demonstrates this behavior:

```python
    import torch
    from time import time

    torch.set_num_threads(1)

    inp = torch.rand((128, 128, 128), device='cpu')
    st = time()
    for _ in range(100):
        _ = torch.std_mean(inp)
    print('cpu std_mean', time() - st)
    inp = torch.rand((128, 128, 128), device='cpu')
    st = time()
    for _ in range(100):
        _ = torch.mean(inp)
        _ = torch.std(inp)
    print('cpu separate', time() - st)
````

> cpu std_mean 0.9533390998840332
> cpu separate 0.17572903633117676

At least on my system, the separate mean + std calls are >5x faster. 

### Versions

Collecting environment information...
PyTorch version: 2.1.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jul  5 2023, 18:54:27) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.5.0-25-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 545.23.08
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             16
On-line CPU(s) list:                0-15
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen 7 5800X3D 8-Core Processor
CPU family:                         25
Model:                              33
Thread(s) per core:                 2
Core(s) per socket:                 8
Socket(s):                          1
Stepping:                           2
Frequency boost:                    enabled
CPU max MHz:                        4548,8281
CPU min MHz:                        2200,0000
BogoMIPS:                           6799.80
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm
Virtualization:                     AMD-V
L1d cache:                          256 KiB (8 instances)
L1i cache:                          256 KiB (8 instances)
L2 cache:                           4 MiB (8 instances)
L3 cache:                           96 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-15
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Mitigation; Safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] adan-pytorch==0.1.0
[pip3] fft-conv-pytorch==1.2.0
[pip3] numpy==1.26.4
[pip3] torch==2.1.1
[pip3] torchaudio==2.1.0.dev20230724
[pip3] torchvision==0.16.0.dev20230724
[pip3] triton==2.1.0
[conda] adan-pytorch              0.1.0                    pypi_0    pypi
[conda] blas                      1.0                         mkl  
[conda] fft-conv-pytorch          1.2.0                    pypi_0    pypi
[conda] mkl                       2023.1.0         h6d00ec8_46342  
[conda] mkl-service               2.4.0           py310h5eee18b_1  
[conda] mkl_fft                   1.3.6           py310h1128e8f_1  
[conda] mkl_random                1.2.2           py310h1128e8f_1  
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] pytorch-cuda              12.1                 ha16c6d3_5    pytorch-nightly
[conda] pytorch-mutex             1.0                        cuda    pytorch-nightly
[conda] torch                     2.1.1                    pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230724     py310_cu121    pytorch-nightly
[conda] torchtriton               2.1.0+9e3e10c5ed           py310    pytorch-nightly
[conda] torchvision               0.16.0.dev20230724     py310_cu121    pytorch-nightly

### PR DESCRIPTION
**Pull request by qflen**
Fixes #122191.

## Summary

```python
torch.set_num_threads(1); x = torch.rand(128, 128, 128)
torch.std_mean(x)              # ~7 ms on M5
torch.std(x); torch.mean(x)    # ~1.8 ms combined
```

`std_var_out` has a CPU all-reduce fast path (added in #43858) that routes fp32/fp64 to a parallel two-pass `(x - mean)^2` in `std_var_all_cpu`. `std_var_mean_out` doesn't, and always dispatches to `std_var_stub` (scalar Welford, no vectorization).

`std_var_all_cpu` already computes the mean internally and throws it away. Returning it and adding the same fast-path branch to `std_var_mean_out` lifts the speedup. Welford stays for per-dim, complex, bf16/fp16, CUDA, and empty reductions. Diff is +13/-4 in `ReduceOps.cpp`.

Note: after this, `std_mean` fp32 all-reduce uses fp32 `cascade_sum` for the mean (as `std` already did) instead of Welford's double accumulator. For pathological `|mean| >> stddev` inputs Welford was marginally more precise; aligning `std_mean` with `std`/`var` is the intended trade-off. Promoting `self.mean()` to double inside `std_var_all_cpu` is a clean follow-up.

## Test plan

Existing `test_std_mean_*` / `test_var_mean_correction` already cross-check fused vs separate. Added `test_std_var_mean_all_reduce_cpu_fast_path` to guard the new branch against an fp64 ground truth.

Apple M5, matched source builds:

| shape | dtype | threads | before | after | speedup |
|---|---|---|---:|---:|---:|
| 1024×1024 | fp32 | 1 | 3372 µs | 842 µs | 4.00× |
| 128×128×128 | fp32 | 1 | 6764 µs | 1685 µs | 4.01× |
| 64×512×512 | fp32 | 1 | 54096 µs | 13759 µs | 3.93× |
| 64×512×512 | fp64 | 1 | 53867 µs | 14625 µs | 3.68× |
| 128×128×128 | fp32 | 10 | 6892 µs | 1727 µs | 3.99× |

Same speedup for `var_mean`. Per-dim, bf16/fp16, and the standalone std/var/mean: within jitter.

### CODE REVIEW

**Review by colesbury — COMMENTED** [review]
This looks okay to me, but it's been a long time since I looked at this code. @albanD does this look okay to you too?

**Comment by qflen** [comment]
Hi, @albanD

Could you review and merge if it looks good?
